src/rust/engine/fs/src/lib.rs

```diff
@@ -729,29 +900,137 @@ pub trait VFS<E: Send + Sync + 'static>: Clone + Send + Sync + 'static {
         expansion
           .todo
           .drain(..)
-          .map(|path_glob| context.expand_single(path_glob, exclude))
+          .map(|sourced_glob| context.expand_single(sourced_glob, exclude))
           .collect::<Vec<_>>()
       });
-      round.map(move |paths_and_globs| {
+      round.map(move |single_expansion_results| {
         // Collect distinct new PathStats and PathGlobs
-        for (paths, globs) in paths_and_globs.into_iter() {
-          expansion.outputs.extend(paths.into_iter().map(|p| (p, ())));
-          let completed = &mut expansion.completed;
+        for exp in single_expansion_results {
+          let SingleExpansionResult {
+            sourced_glob: GlobWithSource { path_glob, source },
+            path_stats,
+            globs,
+          } = exp;
+
+          expansion.outputs.extend(path_stats.clone());
+
           expansion
-            .todo
-            .extend(globs.into_iter().filter(|pg| completed.insert(pg.clone())));
+            .completed
+            .entry(path_glob.clone())
+            .or_insert_with(|| GlobExpansionCacheEntry {
+              globs: globs.clone(),
+              matched: if path_stats.is_empty() {
+                GlobMatch::DidNotMatchAnyFiles
+              } else {
+                GlobMatch::SuccessfullyMatchedSomeFiles
+              },
+              sources: vec![],
+            })
+            .sources
+            .push(source);
+
+          // TODO(cosmicexplorer): is cloning these so many times as `GlobSource`s something we can
+          // bypass by using `Arc`s? Profile first.
+          let source_for_children = GlobSource::ParentGlob(path_glob);
+          for child_glob in globs {
+            if expansion.completed.contains_key(&child_glob) {
```

Comment:
Rather than `contains_key` + `get_mut`, you can use the `Entry` API:

```
if let Occupied(entry) = expansion.completed.entry(&child_glob) {
  entry.get_mut().sources.push(..);
} else {
  ..
}
```

Reply:
I had left a TODO in a previous iteration that this needed to be more ergonomic -- thanks for pointing it out! I've made the change.